### PR TITLE
CloudStack - add diskOffering to machine template

### PR DESCRIPTION
**Pull request by wanyufe**
*Issue #, if available:*
Allow custom to specify diskoffering using ID or name for data disk.
*Description of changes:*
Allow custom to specify diskoffering using ID or name for data disk. mountPath will be used to mount provided data disk.
*Testing (if applicable):*
Unit testing, system testing
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Makefile
Comment (maxdrib) [inline Makefile:342]: why is this change necessary?

File: controllers/controllers/resource/fetcher.go
Comment (maxdrib) [inline controllers/controllers/resource/fetcher.go:620]: don't we need a nil check here for `csMachineTemplate.Spec.Spec.Spec.DiskOffering`?

File: pkg/executables/cmk.go
Comment (maxdrib) [inline pkg/executables/cmk.go:130]: do we need to take "account" or "domain" in this command?

File: Makefile
Comment (wanyufe) [reply Makefile:342]: This is a incompatible find command parameter between linux and macos, my version works for macos. The existed version does not work in my mac. 

Can you also do a test to see if it happens on your mac? You can run make clean to trigger that command.

File: pkg/providers/cloudstack/validator.go
Comment (maxdrib) [inline pkg/providers/cloudstack/validator.go:258]: aren't these required fields already validated the type definition?

**Review by maxdrib — COMMENTED** [review]
minor comments, mostly need more unit tests

File: Makefile
Comment (maxdrib) [reply Makefile:342]: Gotcha, the old version didn't work for me either on Mac, nice catch

File: go.mod
Comment (wanyufe) [reply go.mod:10]: this needs to be updated/changed once CAPC piece being merged and tagged. The version in green is a branch under capc repo.

File: pkg/executables/cmk.go
Comment (wanyufe) [reply pkg/executables/cmk.go:130]: Good question. 
I used our cmk tool to test the list diskofferings and found account was not an option. And if I applied domainId filter, nothing returned, even we do have valid disk offerings under specified domain. So I believe cloudstack's API list diskoffering implementation for domainId/Account is not working.

https://cloudstack.apache.org/api/apidocs-4.14/apis/listDiskOfferings.html

```(localcloud) 🐱 > list diskofferings domainid=5300cdac-74d5-11ec-8696-c81f66d3e965 
(localcloud) 🐱 > list diskofferings 
{
  "count": 4,
  "diskoffering": [
    {
      "created": "2022-01-13T20:01:51-0500",
      "disksize": 5,
      "displayoffering": true,
      "displaytext": "Small Disk, 5 GB",
      "id": "399f1b0d-742f-4474-a138-7932b9a54fed",
      "iscustomized": false,
      "name": "Small",
      "provisioningtype": "thin",
      "storagetype": "shared"
    },
    {
      "created": "2022-01-13T20:01:51-0500",
      "disksize": 20,
      "displayoffering": true,
      "displaytext": "Medium Disk, 20 GB",
      "id": "9c4a12f6-4cc5-470a-98b6-cf40306f0e51",
      "iscustomized": false,
      "name": "Medium",
      "provisioningtype": "thin",
      "storagetype": "shared"
    },
    {
      "created": "2022-01-13T20:01:51-0500",
      "disksize": 100,
      "displayoffering": true,
      "displaytext": "Large Disk, 100 GB",
      "id": "3a9f4649-67e3-4af8-8a9c-37a582383217",
      "iscustomized": false,
      "name": "Large",
      "provisioningtype": "thin",
      "storagetype": "shared"
    },
    {
      "created": "2022-01-13T20:01:51-0500",
      "disksize": 0,
      "displayoffering": true,
      "displaytext": "Custom Disk",
      "id": "df36c956-9f4a-4cc3-84c9-239e0914d0f8",
      "iscustomized": true,
      "name": "Custom",
      "provisioningtype": "thin",
      "storagetype": "shared"
    }
  ]
}
(localcloud) 🐱 > list diskofferings 
domainid=    filter=      id=          isrecursive= keyword=     listall=     name=        page=        pagesize=    zoneid=  
```

File: pkg/providers/cloudstack/validator.go
Comment (wanyufe) [reply pkg/providers/cloudstack/validator.go:258]: If you provided the required fields with an empty string in the yaml file, it can pass the type definition and I think this validator is still needed.

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:77]: more unit testing cases added for different fields of diskoffering struct.

File: go.mod
Comment (wanyufe) [reply go.mod:10]: agree.

**Review by maxdrib — APPROVED** [review]

File: Makefile
Comment (wanyufe) [reply Makefile]: Updated with conditional statement to work for both MacOs and Linux

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:73]: can we use `ValidatePath`?

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:80]: can we reuse the validation methods in api pkg?

File: pkg/executables/cmk.go
Comment (jiayiwang7) [inline pkg/executables/cmk.go:143]:
```suggestion
		return fmt.Errorf("parsing response into json: %v", err)
```

File: pkg/providers/cloudstack/validator.go
Comment (jiayiwang7) [inline pkg/providers/cloudstack/validator.go:248]: this validation seems redundant since it's already done in API context unware validation.

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_types.go:54]: We (eksa, cloudstack) don't have a limit on what file systems allowed. The limitation comes from the supported fs in mkfs tool in used image. 

If we have an allowed list, we need to keep it in sync with supported file system in customer's image. 
The benefits of having an allowed list is to pre-flight check some typo, the disadvantage is that we have to keep it current/right which is extra ongoing work.

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:80]: We can re-use the validatePath like before comments. But the current validate method implementation in api pkg is not detailed enough to tell which field is invalid. 

Let me refactor the api validate method to combine validatePath and other fields together and return detailed info so that it can re-used.

File: pkg/providers/cloudstack/validator.go
Comment (wanyufe) [reply pkg/providers/cloudstack/validator.go]: refactored  diskoffering api validation so that it can be used else.

File: pkg/executables/cmk.go
Comment (wanyufe) [reply pkg/executables/cmk.go:143]: all other places while checking json response in file cmk.go use 'failed to parse ...'. 

Any particular reason to use 'parsing ...' ?

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go]: refactored and reused.

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go]: Updated.

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_types.go:50]: add extra comments

File: pkg/executables/cmk.go
Comment (jiayiwang7) [reply pkg/executables/cmk.go:143]: rather redundant for a type error to say "failed to ....", we recently updated all our errors to remove it. So yea pls do update them all.

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cloudstackmachineconfig_types.go:77]: This method currently is only being called in webhook (controller). We need it in the config manager / API validation as well for CLI validation. Can we add it in https://github.com/aws/eks-anywhere/blob/main/pkg/api/v1alpha1/cloudstackmachineconfig_types.go#L98
Also curious where the other context unaware cloudstack machine config API validation are?

Comment (jiayiwang7) [reply pkg/api/v1alpha1/cloudstackmachineconfig_types.go:77]: The return should only include `error`, you don't need to include the key/value in webhook error

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cloudstackmachineconfig_types.go:93]: why sum(len) > 0 means empty id/name?

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:40]:
```suggestion
		return apierrors.NewBadRequest("disk offering [a:b], preventing CloudStackMachineConfig resource creation")
```

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:68]: same here, no need to have the fieldName and value as return

File: pkg/providers/cloudstack/validator.go
Comment (jiayiwang7) [inline pkg/providers/cloudstack/validator.go:248]: this validation should be moved to config manager validation I mentioned before since it's context unaware.

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (jiayiwang7) [inline pkg/api/v1alpha1/cloudstackmachineconfig_types.go:99]: do we need this? :0

**Review by maxdrib — APPROVED** [review]
looks like there are some linting issues in the PR but I don’t have any comments except for what Joey provided

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_types.go:93]: This `else` section is try to catch a situation that no id AND no name defined, but at least one of mountPath, filesystem, device and label defined. 

If every thing is empty (id/name/mountPath/filesystem/device/label), it is valid and being considered as disk offering not provided.

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:40]: discussed with Jia offline and reached an agreement that field name and value are needed for validateUpdate in webhook call. 

In validateCreate, a simple bad request can be returned, but for validteUpdate, field name has to be specified in exception.

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_types.go:77]: discussed with Jia offline and reached an agreement that field name and value are needed for validateUpdate in webhook call.

In validateCreate, a simple bad request can be returned, but for validteUpdate, field name has to be specified in exception.

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:68]: discussed with Jia offline and reached an agreement that field name and value are needed for validateUpdate in webhook call.

In validateCreate, a simple bad request can be returned, but for validteUpdate, field name has to be specified in exception.

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (wanyufe) [reply pkg/api/v1alpha1/cloudstackmachineconfig_types.go:99]: trying to re-use this logic in validateMachineConfig in validator.go, so this simple method is created.

File: pkg/providers/cloudstack/validator.go
Comment (wanyufe) [reply pkg/providers/cloudstack/validator.go]: The validation in this commit is static validation which means it does not need extra info to do the validation and by design it should be under cloudstackmachineconfig_types.go's validate method. That validate method is triggered by config manager's reflection call so that it can be applied to every provider (vsphere/cloudstack, etc) that implements validate interface.

The original code is a TODO comment https://github.com/aws/eks-anywhere/blob/main/pkg/providers/cloudstack/validator.go#L124

I think this refactor is out of scope for this PR, and I am creating an issue to address diskoffering validate, service offering, compute offering and user validation. It seems at least vsphere and cloudstack needs these kind of refactor.

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go
Comment (jiayiwang7) [reply pkg/api/v1alpha1/cloudstackmachineconfig_types.go:99]: I feel this method name is confusing. What does it mean by Provided? I need to go back and see `len(r.Id) > 0 || len(r.Name) > 0`. imo it's more clear just to use the if statement. But your call

File: pkg/providers/cloudstack/validator.go
Comment (jiayiwang7) [inline pkg/providers/cloudstack/validator.go:203]: this should be part of config manager validation

File: pkg/executables/cmk.go
Comment (wanyufe) [reply pkg/executables/cmk.go:143]: update all `failed to parse' in cmk.go to `parsing`

File: pkg/providers/cloudstack/validator.go
Comment (wanyufe) [reply pkg/providers/cloudstack/validator.go:203]: There will be a follow-up PR to move static validation away from validator.go

**Pull request merged**
